### PR TITLE
`Pool` struct, custom providers

### PR DESCRIPTION
**Pull request by ts0yu**
- this is a pr for a gradual refactor i am working on
- will include custom providers, generic data sources and traits and impls for most files

### CODE REVIEW

**Comment by ts0yu** [comment]
tests don't pass yet

**Comment by ts0yu** [comment]
i decided to use env variables instead of cli, i know there was discussion in [16](https://github.com/primitivefinance/arbiter/issues/16)

**Comment by Autoparallel** [comment]
> i decided to use env variables instead of cli, i know there was discussion in [16](https://github.com/primitivefinance/arbiter/issues/16)

That's fine! If need be we can always change inputs.

File: README.md [hunk 1]
```diff
@@ -46,15 +46,15 @@ USAGE:
...
+        --api_key     Provide an Etherscan API key if function calls require it (e.g., providing a token that is not in
+                      the database).
+    -h, --help        Prints help information
+    -V, --version     Prints version information
 
 OPTIONS:
         --fee <fee>          Specifies the basis points for the pool. [1, 5, 30, 100] [default: 5]
         --token0 <token0>    Specifies the first token for a token pair. [default: ETH]
-        --token1 <token1>    Specifies the second token for a token pair which will be the numeraire. [default: USDC]
+        --token1 <token1>    Specifies the second token for a token pair which we be the numeraire. [default: USDC]
```
Comment (Autoparallel) [inline README.md:57]: typo?

Comment (ts0yu) [reply README.md:57]: oops yeah, let me fix alongside some fmt stuff to make it pass the ci

File: app/src/tokens/mod.rs [hunk 2]
```diff
@@ -1,191 +1,145 @@
...
     let tether_address = "0xdAC17F958D2ee523a2206206994597C13D831ec7"
         .parse::<Address>()
         .unwrap();
-    let tether = Token {
-        address: tether_address,
-        decimals: 6,
-        name: "USDT".to_string(),
-        is_stable: true,
-    };
+    let tether = Token::new(tether_address, 6, String::from("USDT"), true);
```
Comment (Autoparallel) [inline app/src/tokens/mod.rs:56]: Naming convention differs here

**Review by 0xJepsen — CHANGES_REQUESTED** [review]
This looks great! We will merge it after Colin's comments are addressed. I have some ideas for a future direction that could be good.

**Review by Autoparallel — COMMENTED** [review]
I left a few comments. Nothing big.

Thoughts on token naming scheme? Things that came up for me:
`tether` (`tokens/mod.rs` line 52)
`s_eth2` (`tokens/mod.rs` line 85)

File: app/src/tokens/mod.rs (hunk 2 above)
Comment (ts0yu) [reply app/src/tokens/mod.rs]: fixed

**Review by 0xJepsen — APPROVED** [review]

**Pull request merged**